File: core/config/saver.go
// Package config provides configuration management for elmos.
// This file contains configuration saving and profile management.
package config

import (
	"fmt"
	"os"
	"path/filepath"

	"gopkg.in/yaml.v3"
)

// saveConfig is the serialised form of the workspace configuration.
// Only the fields that are written to elmos.yaml are included here;
// computed/runtime fields (Machines, CurrentMachine, Plugins, etc.) are omitted.
type saveConfig struct {
	Image    ImageConfig              `yaml:"image"`
	Build    BuildConfig              `yaml:"build"`
	QEMU     QEMUConfig               `yaml:"qemu"`
	Paths    PathsConfig              `yaml:"paths,omitempty"`
	Profiles map[string]ProfileConfig `yaml:"profiles,omitempty"`
}

// Save saves the configuration to a YAML file.
// Keys in the output match the mapstructure tags so the file can be
// loaded back by config.Load without any value loss.
func (cfg *Config) Save(path string) error {
	defaults := cfg.computeDefaults()
	prepared := cfg.prepareForSave(defaults)

	sc := saveConfig{
		Image:    prepared.Image,
		Build:    prepared.Build,
		QEMU:     prepared.QEMU,
		Paths:    prepared.Paths,
		Profiles: prepared.Profiles,
	}

	data, err := yaml.Marshal(sc)
	if err != nil {
		return fmt.Errorf("failed to marshal config: %w", err)
	}

	if err := ensureDir(filepath.Dir(path)); err != nil {
		return err
	}

	if err := os.WriteFile(path, data, 0644); err != nil {
		return fmt.Errorf("failed to write config file: %w", err)
	}

	return nil
}

// computeDefaults calculates default config for comparison.
func (cfg *Config) computeDefaults() *Config {
	defaults := &Config{}
	if cfg.Paths.ProjectRoot != "" {
		defaults.Paths.ProjectRoot = cfg.Paths.ProjectRoot
	}
	applyComputedDefaults(defaults)
	return defaults
}

// prepareForSave creates a copy with default values cleared.
func (cfg *Config) prepareForSave(defaults *Config) Config {
	saveCfg := *cfg
	saveCfg.Paths = clearDefaultPaths(cfg.Paths, defaults.Paths)
	saveCfg.Image = clearDefaultImage(cfg.Image, defaults.Image)
	return saveCfg
}

// clearDefaultPaths clears path values that match defaults.
// ProjectRoot is ALWAYS saved so paths resolve correctly even when the
// config is loaded from a different working directory (e.g. via --config).
func clearDefaultPaths(paths, defaults PathsConfig) PathsConfig {
	result := paths
	// ProjectRoot is intentionally never cleared — it anchors all
	// workspace-relative path computations.
	if paths.KernelDir == defaults.KernelDir {
		result.KernelDir = ""
	}
	if paths.ModulesDir == defaults.ModulesDir {
		result.ModulesDir = ""
	}
	if paths.AppsDir == defaults.AppsDir {
		result.AppsDir = ""
	}
	if paths.LibrariesDir == defaults.LibrariesDir {
		result.LibrariesDir = ""
	}
	if paths.PatchesDir == defaults.PatchesDir {
		result.PatchesDir = ""
	}
	if paths.RootfsDir == defaults.RootfsDir {
		result.RootfsDir = ""
	}
	if paths.DiskImage == defaults.DiskImage {
		result.DiskImage = ""
	}
	return result
}

// clearDefaultImage clears image values that match defaults.
func clearDefaultImage(image, defaults ImageConfig) ImageConfig {
	result := image
	if image.Path == defaults.Path {
		result.Path = ""
	}
	if image.MountPoint == defaults.MountPoint {
		result.MountPoint = ""
	}
	return result
}

// ensureDir creates a directory if it doesn't exist.
func ensureDir(dir string) error {
	if err := os.MkdirAll(dir, 0755); err != nil {
		return fmt.Errorf("failed to create config directory %s: %w", dir, err)
	}
	return nil
}

// ApplyProfile applies a named profile to the current configuration.
func (cfg *Config) ApplyProfile(name string) error {
	profile, ok := cfg.Profiles[name]
	if !ok {
		return fmt.Errorf("profile not found: %s", name)
	}

	if profile.Arch != "" {
		cfg.Build.Arch = profile.Arch
	}
	if profile.Jobs > 0 {
		cfg.Build.Jobs = profile.Jobs
	}
	if profile.Memory != "" {
		cfg.QEMU.Memory = profile.Memory
	}
	if profile.CrossCompile != "" {
		cfg.Build.CrossCompile = profile.CrossCompile
	}

	return nil
}

// GetArchConfig returns the architecture configuration for the current build arch.
func (cfg *Config) GetArchConfig() *ArchConfig {
	return GetArchConfig(cfg.Build.Arch)
}

File: src/config/saver.cpp
// ============================================================================
// config/saver.cpp — Configuration saving and profile management
// Replaces Go's core/config/saver.go
// ============================================================================

#include "arch.hpp"
#include "defaults.hpp"
#include "machine.hpp"
#include "types.hpp"

#include <yaml-cpp/yaml.h>

#include <filesystem>
#include <fstream>

namespace elmos::config {

auto Config::get_machine(const std::string& name) -> MachineDefinition* {
    auto it = machines.find(name);
    if (it == machines.end())
        return nullptr;
    return it->second.get();
}

auto Config::set_current_machine(std::shared_ptr<MachineDefinition> m) -> VoidResult {
    if (!m)
        return make_error(Error::config("machine definition cannot be nil"));
    auto r = m->validate();
    if (!r)
        return r;
    current_machine = std::move(m);
    build.arch = current_machine->kernel.arch;
    machine = current_machine->name;
    return {};
}

auto Config::apply_profile(const std::string& name) -> VoidResult {
    auto it = profiles.find(name);
    if (it == profiles.end()) {
        return make_error(Error::config("profile not found: " + name));
    }
    const auto& profile = it->second;
    if (!profile.arch.empty())
        build.arch = profile.arch;
    if (profile.jobs > 0)
        build.jobs = profile.jobs;
    if (!profile.memory.empty())
        qemu.memory = profile.memory;
    if (!profile.cross_compile.empty())
        build.cross_compile = profile.cross_compile;
    return {};
}

auto Config::get_arch_config() const -> const ArchConfig* {
    return config::get_arch_config(build.arch);
}

auto Config::save(const std::string& path) -> VoidResult {
    YAML::Emitter out;
    out << YAML::BeginMap;

    // Image
    out << YAML::Key << "image" << YAML::Value << YAML::BeginMap;
    if (!image.path.empty())
        out << YAML::Key << "path" << YAML::Value << image.path;
    if (!image.volume_name.empty())
        out << YAML::Key << "volume_name" << YAML::Value << image.volume_name;
    if (!image.size.empty())
        out << YAML::Key << "size" << YAML::Value << image.size;
    if (!image.mount_point.empty())
        out << YAML::Key << "mount_point" << YAML::Value << image.mount_point;
    out << YAML::EndMap;

    // Build
    out << YAML::Key << "build" << YAML::Value << YAML::BeginMap;
    out << YAML::Key << "arch" << YAML::Value << build.arch;
    out << YAML::Key << "jobs" << YAML::Value << build.jobs;
    out << YAML::Key << "llvm" << YAML::Value << build.llvm;
    if (!build.cross_compile.empty())
        out << YAML::Key << "cross_compile" << YAML::Value << build.cross_compile;
    out << YAML::Key << "verbose" << YAML::Value << build.verbose;
    out << YAML::EndMap;

    // QEMU
    out << YAML::Key << "qemu" << YAML::Value << YAML::BeginMap;
    out << YAML::Key << "memory" << YAML::Value << qemu.memory;
    out << YAML::Key << "gdb_port" << YAML::Value << qemu.gdb_port;
    out << YAML::Key << "ssh_port" << YAML::Value << qemu.ssh_port;
    out << YAML::Key << "smp" << YAML::Value << qemu.smp;
    out << YAML::EndMap;

    // Paths (only non-default values)
    out << YAML::Key << "paths" << YAML::Value << YAML::BeginMap;
    if (!paths.project_root.empty())
        out << YAML::Key << "project_root" << YAML::Value << paths.project_root;
    out << YAML::EndMap;

    out << YAML::EndMap;

    // Write to file
    namespace fs = std::filesystem;
    if (auto parent = fs::path(path).parent_path(); !parent.empty()) {
        std::error_code ec;
        fs::create_directories(parent, ec);
        if (ec) {
            return make_error(Error::config("cannot create config directory: " + ec.message()));
        }
    }

    std::ofstream file(path, std::ios::out | std::ios::trunc);
    if (!file) {
        return make_error(Error::config("cannot write config file: " + path));
    }
    file << out.c_str();
    return {};
}

}  // namespace elmos::config
